File: algo.py
"""
complexity of this algo O(10^length of unique string)
for example: In the case of Send,More and Money
The Unique string will be SENDMORY,its length is 8,so the complexity is O(10^8)
"""
class cryp:
    def __init__(self) -> None:
        self.flag = False
        self.resList = []
    def makenum(self,map:dict,str:str):
        s = ""
        for i in str:
            s = s + map[i].__str__()
        return int(s)
    def checkForInitialZeroes(self,words:list,map:dict):
        for word in words:
            if(map[word[0]]==0):
                return True
        return False
    def helper(self,map:dict,unique:str,used:list,words:list,result:str,target:int)->None:
        # base case
        if(target==len(unique)):
            if(map[result[0]]==0 or self.checkForInitialZeroes(words,map)):
                return
            sum=0
            for i in words:
                sum = sum + self.makenum(map,i)
            res = self.makenum(map,result)
            if(sum==res):
                self.flag = True
                ans=[]
                for i in map:
                    ans.append(i + ": " +map[i].__str__())
                self.resList.append(ans)
            return
        ch = unique[target]
        for i in range(0,10):
            if(used[i]==False):
                map[ch] = i
                used[i] = True
                self.helper(map,unique,used,words,result,target+1)
                used[i] = False
    def isSolvable(self,words:list,result:str)->bool:
        map = {}
        unique = ""
        for i in words:
            for j in i:
                if((j in map)==False):
                    map[j] = -1
                    unique = unique+j
        for i in result:
            if((i in map)==False):
                map[i] = -1
                unique =unique + i
        print(unique)
        if(len(unique)>10):
            return False
        used=[]
        for i in range(0,10):
            used.append(False)
        self.helper(map,unique,used,words,result,target=0)
        print(len(map))
        return self.flag
if __name__=='__main__':
    words = ["ABCD","EF"]
    res = "ABCG"
    ob = cryp()
    print(ob.isSolvable(words,res))
    print(ob.resList)
    print(len(ob.resList))
        
                        

File: algo.java

import java.util.*;
public class algo {
    static boolean flag = false;
    public static int makenum(HashMap<Character,Integer> map,String str){
        StringBuilder sb = new StringBuilder();
        for(int i=0;i<str.length();i++){
            sb.append(map.get(str.charAt(i)));
        }
        String num = sb.toString();
        return (Integer.parseInt(num));
    }
    public static void helper(HashMap<Character,Integer> map,String unique,boolean[] used,String[] words,String result,int target){
        //base case
        if(target==unique.length()){
            int sum = 0;
            for(int i=0;i<words.length;i++){
                sum = sum + makenum(map,words[i]);
            }
            int res = makenum(map,result);
            if(sum==res){
                System.out.println("SUM:"+sum);
                flag=true;
                System.out.println(map);
            }
            return;
        }
        char ch = unique.charAt(target);
        for(int i=0;i<=9;i++){
            if(used[i]==false){
                map.put(ch,i);
                used[i] = true;
                helper(map,unique,used,words,result,target+1);
                //bactracking step (undo changes)
                used[i] = false;
            }
        }
    }
    public static boolean isSolvable(String[] words, String result) {
        HashMap<Character,Integer> map = new HashMap<>();
        StringBuilder sb = new StringBuilder();
        for(int i=0;i<words.length;i++){
            for(int j=0;j<words[i].length();j++){
                char ch = words[i].charAt(j);
                if(map.containsKey(ch)==false){
                    map.put(ch,-1);
                    sb.append(ch);
                }
            }
        }
        for(int i=0;i<result.length();i++){
            char ch = result.charAt(i);
            if(map.containsKey(ch)==false){
                    map.put(ch,-1);
                    sb.append(ch);
                }
        }
        String unique = sb.toString();
        System.out.println("Unique: "+unique);
        if(unique.length()>10) return false;
        boolean used[] = new boolean[10];
        System.out.println(Arrays.toString(used));
        helper(map,unique,used,words,result,0);
        System.out.println(map.size());
        return flag;
    }
    public static void main(String[] args) {
        String words[] = {"SEND","MORE"};
        String res = "MONEY";
        System.out.println(isSolvable(words, res));
    }
}